### PR TITLE
SI-7669 Fix exhaustivity warnings for recursive ADTs.

### PR DESCRIPTION
**Pull request by retronym**
The pattern matcher's analysis was correctly finding
models under which the match in the enclosed test
could fail. But, when trying to render that model
as a counter example, it ran into an internal inconsistency
and gave up.

That inconsistency arose from VariableAssignment, which:

> turn the variable assignments into a tree
> the root is the scrutinee (x1), edges are labelled
> by the fields that are assigned a node is a variable
> example (which is later turned into a counter example)

In the process, it notes the unreachable case `V2 = NotHandled`,
which can only arise if `V1 = Op`, ie the scrutinee is `Op(NotHandled())`.
V2 is assosicated with the path `x1.arg`. The code then looked for
any variable assosicated with the prefix `x1` and registered that its
field `arg` was assosicated with this variable assignment.

However, the assignment for `V1 = NotHandled` (another missing case)
is also associated with the path `x1`. Registering this field makes
no sense here; we should only do that for `Op`.

This commit conditioanlly registers the fields based on the class
of `VariableAssignment#cls`. We no longer hit the inconsistency in
`VariableAssignment#allFieldAssignmentsLegal`.

This changes the results of an existing test, in what seems to be
a positive way.

Review by @adriaanm

### CODE REVIEW

**Comment by retronym** [comment]
@adriaanm I've introduced some non-determinism into the warnings; that's what's breaking the build:

```
ticket/7669 ~/code/scala qbin/scalac test/files/neg/exhausting.scala 2>&1 | grep --context=2 :21
test/files/neg/exhausting.scala:21: warning: match may not be exhaustive.
It would fail on the following inputs: List(_), List(_, _, _)
  def fail1[T](xs: List[T]) = xs match {
ticket/7669 ~/code/scala qbin/scalac test/files/neg/exhausting.scala 2>&1 | grep --context=2 :21
test/files/neg/exhausting.scala:21: warning: match may not be exhaustive.
It would fail on the following inputs: List(_), List(_, _, _)
  def fail1[T](xs: List[T]) = xs match {
ticket/7669 ~/code/scala qbin/scalac test/files/neg/exhausting.scala 2>&1 | grep --context=2 :21
test/files/neg/exhausting.scala:21: warning: match may not be exhaustive.
It would fail on the following input: List(_, _, _)
  def fail1[T](xs: List[T]) = xs match {
```

Any ideas what might trigger that?

**Comment by retronym** [comment]
I've stemmed the non-determinism by canonicalizing `Nil` as `List()` in `toCounterExample`. Before, having both in the list of counter examples could lead to pruning out both of them. It seems to be ordering dependent, and we must have some upstream non-determinism in the model generation that highlighted this problem. It might be worth weeding that out, too.

**Comment by retronym** [comment]
On second thoughts, we probably should rewrite `prune` to avoid using `==` (ie, avoid using `distinct`), in favour of just using `coveredBy`. Just run through the examples and add them to the result set if not covered by any of the result set. It's late here so I'll try that tomorrow...

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 1]
```diff
@@ -622,6 +632,11 @@ trait MatchAnalysis extends MatchApproximation {
         private lazy val cls        = if (ctor == NoSymbol) NoSymbol else ctor.owner
         private lazy val caseFieldAccs = if (cls == NoSymbol) Nil else cls.caseFieldAccessors
 
+        def addField(symbol: Symbol, assign: VariableAssignment) {
+          // SI-7669 Only register this field if if this class contains it.
+          val isInvalidField = symbol.isCaseAccessor && !caseFieldAccs.contains(symbol)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: Is the variable name off by "In-"?

**Comment by adriaanm** [comment]
> On second thoughts, we probably should rewrite prune to avoid using == (ie, avoid using distinct), in favour of just using coveredBy. Just run through the examples and add them to the result set if not covered by any of the result set. It's late here so I'll try that tomorrow...

sounds good!

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: Nope, we're checking if its a field that _doesn't_ exist in this type so it is _in_ valid.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: Phrase it positively and rechristen as `canConstrainField`?

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: done

**Comment by retronym** [comment]
> On second thoughts, we probably should rewrite prune to avoid using == (ie, avoid using distinct), in favour of just using coveredBy. Just run through the examples and add them to the result set if not covered by any of the result set. It's late here so I'll try that tomorrow...

done now

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 2]
```diff
@@ -488,8 +488,15 @@ trait MatchAnalysis extends MatchApproximation {
...
     object CounterExample {
       def prune(examples: List[CounterExample]): List[CounterExample] = {
-        val distinct = examples.filterNot(_ == NoExample).toSet
-        distinct.filterNot(ce => distinct.exists(other => (ce ne other) && ce.coveredBy(other))).toList
+        // SI-7669 Warning: we don't used examples.distinct here any more as
+        //         we can have A != B && A.coveredBy(B) && B.coveredBy(A)
+        //         with Nil and List().
+        val result = mutable.Buffer[CounterExample]()
+        for (example <- examples) {
+          if (!result.exists(other => example.coveredBy(other)))
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: I prefer `example coveredBy _`, but that's just nitpicking.

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: consider that nit picked.

**Comment by gkossakowski** [comment]
@retronym: what's the status of this PR?

**Comment by retronym** [comment]
I believe that I've addressed the feedback, ready for re-review by @adriaanm

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**